### DISCUSSION TITLE
GPU fusion kernel perf drops when it has different layouts for the I/O tensors.

### DISCUSSION THREAD

**Discussion by shawnwang18** [openxla/xla#766] [General]
<html>
<body>
<b style="font-weight:normal;" id="docs-internal-guid-1ef47a05-7fff-d787-1906-8c199bd8d9a9"><p dir="ltr" style="line-height:1.38;margin-top:0pt;margin-bottom:0pt;"><span style="font-size:11pt;font-family:Arial;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">We observed that for XLA fusion kernel, performance may severely regress when the fusion kernel uses different layouts for its input/output tensors. The issue looks like frequently occurred when multiple layouts are used in the same XLA model.</span></p><br /><p dir="ltr" style="line-height:1.38;margin-top:0pt;margin-bottom:0pt;"><span style="font-size:11pt;font-family:Arial;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">In principle, we want the read/write of the fusion kernel I/O tensors to be coalesced, so memory access can benefit from cache locality. Memory access coalescence is usually guaranteed when fusion kernel&rsquo;s I/O layout is the same, for example, in kLoop fused kernel, neighbor GPU threads are processing output elements whose addresses are consecutive, and the input data elements kernel needs to fetch are also from consecutive memory address across neighbor GPU threads, if both input/output tensors are using the same memory layout.&nbsp;</span></p><br /><p dir="ltr" style="line-height:1.38;margin-top:0pt;margin-bottom:0pt;"><span style="font-size:11pt;font-family:Arial;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">While we see that multiple XLA fusion optimization passes do not consider the tensor layout impact, which ends up with generating a super large fusion kernel which shows very bad data locality.&nbsp; An simple example which generates different I/O tensor layouts is like below:&nbsp;&nbsp;</span></p><p dir="ltr" style="line-height:1.38;margin-top:0pt;margin-bottom:0pt;"><span style="font-size:11pt;font-family:Arial;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;"><span style="border:none;display:inline-block;overflow:hidden;width:555px;height:480px;"><img src="https://lh6.googleusercontent.com/jwjZak-xHLmwcaB-AnFcBpCKk2f1eAXeOE7mNtZOzJmJaFkGjczu9UOj3x0IQys7dNAxmMGpPihakLEep-nZiu_395c3nX48-n8NO4VfwdjMz6Rmztyu3tgZ5dPPBsWPTURnN6ItBXlFWAYbPO_V0Ah-pehBIFCqd0C_BxRCsJWAIm5Dz-fjmRhS9hKbkQ" width="555" height="480" style="margin-left:0px;margin-top:0px;" /></span></span></p><br /><p dir="ltr" style="line-height:1.38;margin-top:0pt;margin-bottom:0pt;"><span style="font-size:11pt;font-family:Arial;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">The above graph shows the layout after GpuLayoutAssignment pass, here the custom-call instruction requires mandatory layout {1,0,2,3}, and Parameter instruction also requires mandatory layout {3,2,1,0}.&nbsp; XLA will introduce the copy operator to transform the layouts, as the Add operator requires its operand layout to be the same. During GpuInstructionFusion pass, the Add and Copy operator will be fused, this will create a fusion kernel where its operand tensors with different layout, and the locality is bad when reading the Parameter instruction&rsquo;s output tensor inside the fused kernel.&nbsp;</span></p><br /><p dir="ltr" style="line-height:1.38;margin-top:0pt;margin-bottom:0pt;"><span style="font-size:11pt;font-family:Arial;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">&nbsp;What makes the perf even worse is that later fusion pass, like FusionMerger, GpuMultiOutputFusion, GpuHorizontalInputFusion is also not considering any layout impact, which ends up generating a super large fusion kernel with different layout I/O tensors in some cases.&nbsp;</span></p><br /><p dir="ltr" style="line-height:1.38;margin-top:0pt;margin-bottom:0pt;"><span style="font-size:11pt;font-family:Arial;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">We have a case in stable diffusion benchmark, where the fused kernel has different I/O layouts, if we compare with the same fusion pattern but using the same layout, current fusion kernel&rsquo;s perf may be 9x slower.&nbsp;</span></p><div dir="ltr" style="margin-left:0pt;" align="left">

Fused computation | I/O tensors with different layouts | I/O tensors use the same layout {1,0,2,3} | I/O tensors use the same layout {3,2,1,0}
-- | -- | -- | --
fused_computation.1116 | 21.63ms | 2.43ms | 2.43ms

</div><br /><p dir="ltr" style="line-height:1.38;margin-top:0pt;margin-bottom:0pt;"><span style="font-size:11pt;font-family:Arial;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">The unittest and before/after optimization HLO graph can be accessed </span><a href="https://drive.google.com/drive/folders/1BRA4BExE_v0xIOm5v-nRf_Xkb6gDoqhH?usp=share_link" style="text-decoration:none;"><span style="font-size:11pt;font-family:Arial;color:#1155cc;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:underline;-webkit-text-decoration-skip:none;text-decoration-skip-ink:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">here</span></a><span style="font-size:11pt;font-family:Arial;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">.&nbsp;</span></p><br /><p dir="ltr" style="line-height:1.38;margin-top:0pt;margin-bottom:0pt;"><span style="font-size:11pt;font-family:Arial;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">We (Nvidia) had an internal discussion about the issue, possible idea is to resolve the issue like:</span></p><ol style="margin-top:0;margin-bottom:0;padding-inline-start:48px;"><li dir="ltr" style="list-style-type:decimal;font-size:11pt;font-family:Arial;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;" aria-level="1"><p dir="ltr" style="line-height:1.38;margin-top:0pt;margin-bottom:0pt;" role="presentation"><span style="font-size:11pt;font-family:Arial;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">When checking fusing of Copy/Transpose instruction, we check whether this will bring too much uncoalesced access, if so, not fusing it. Also needs to add constraints about layouts in other fusion passes (FusionMerge, GpuMultiOutputFusion, GpuHorizontalInputFusion).</span></p></li><li dir="ltr" style="list-style-type:decimal;font-size:11pt;font-family:Arial;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;" aria-level="1"><p dir="ltr" style="line-height:1.38;margin-top:0pt;margin-bottom:0pt;" role="presentation"><span style="font-size:11pt;font-family:Arial;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">If we fuse the Copy/Transpose operator, is it possible to use tiled transposed algorithm when codegen the kernel, a similar approach used by </span><a href="https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc#L1782" style="text-decoration:none;"><span style="font-size:11pt;font-family:'Courier New';color:#1155cc;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:underline;-webkit-text-decoration-skip:none;text-decoration-skip-ink:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">IrEmitterUnnested::EmitUnnestedTranspose</span></a></p></li></ol><br /><p dir="ltr" style="line-height:1.38;margin-top:0pt;margin-bottom:0pt;"><span style="font-size:11pt;font-family:Arial;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">We would like to initiate the discussion on the possible fix for the issue, any idea?&nbsp;</span></p><br /><br /></b>
</body>
</html>

**Comment by cheshire** [comment]
>  and the locality is bad when reading the Parameter instruction’s output tensor inside the fused kernel. 

I think that's the bug right there: with proper codegen we should never have uncoalesced reads inside the kernel.

This is already covered by your point (2): all fused kTranspose instructions should go through transpose codegen, and if they don't, it's a bug. Could you post an example of the transpose instruction, after all optimizations, which does not?
